lib/elixir/pages/anti-patterns/code-anti-patterns.md

```diff
@@ -319,7 +319,68 @@ This anti-pattern was formerly known as [Accessing non-existent Map/Struct field
 
 ## Dynamic atom creation
 
-TODO.
+#### Problem
+
+An `Atom` is a basic data type of Elixir whose value is its own name. *Atoms* are often useful to identify resources or express the state or result of an operation. Creatings *atoms* dynamically is not an anti-pattern by itself; however, *atoms* are not collected by the BEAM's garbage collector, so values of this type live in memory while software is executing, during its entire lifetime. Also, BEAM limits the number of *atoms* that can exist in an application (*1_048_576* by default) and each *atom* has a maximum size limited to 255 Unicode code points. For these reasons, creating *atoms* dynamically is considered an anti-pattern, since in this way the developer has no control over how many *atoms* will be created during the software execution. This unpredictable scenario can expose the software to unexpected behavior caused by excessive memory usage, or even by reaching the maximum number of *atoms* possible.
+
+#### Example
+
+Imagine that you are implementing a code that performs the conversion of *string* values into *atoms* to identify resources. These *strings* can come from user input or even have been received as response from requests to an API. As this is a dynamic and unpredictable scenario, it is possible for identical *strings* to be converted into new *atoms* that are repeated unnecessarily. This kind of conversion, in addition to wasting memory, can be problematic for the software if it happens too often.
+
+```elixir
+defmodule Identifier do
+  def generate(id) when is_binary(id) do
+    String.to_atom(id)  #<= dynamic atom creation!!
+  end
+end
+```
+
+```elixir
+iex> string_from_user_input = "my_id"
+"my_id"
+iex> string_from_api_response = "my_id"
+"my_id"
+iex> Identifier.generate(string_from_user_input)
+:my_id
+iex> Identifier.generate(string_from_API_response)
+:my_id   #<= atom repeated was created!
+```
+
+When we use the `String.to_atom/1` function to dynamically create an *atom*, it is created regardless of whether there is already another one with the same value in memory, so when this happens automatically, we will not have control over meeting the limits established by BEAM.
+
+#### Refactoring
+
+To remove this anti-pattern, we must ensure that all the identifier *atoms* are created statically, only once, at the beginning of an application's execution:
+
+```elixir
+# statically created atoms...
+_ = :my_id
+_ = :my_id2
+_ = :my_id3
+_ = :my_id4
+```
```

Comment:
I've never thought about doing this, what would be the right place to do this? `application.ex`?

Reply:
This is not a common pattern in my experience though.

Reply:
This advice is not correct as well. The atoms must be defined on the module that attempts to call `String.to_existing_atom`, because code is loaded at module-level, not at an application level. The notes here are up to date: https://hexdocs.pm/elixir/String.html#to_existing_atom/1 :)